cli/bench/space.go

```diff
@@ -0,0 +1,52 @@
+package bench
+
+import (
+	"fmt"
+	"reflect"
+
+	"github.com/FZambia/tarantool"
+	"github.com/tarantool/cartridge-cli/cli/context"
+)
+
+func CreateSpace(tarantoolConnection *tarantool.Connection) error {
+	createCommand := "return box.schema.space.create(...).name"
+	_, err := tarantoolConnection.Exec(tarantool.Eval(createCommand, []interface{}{benchSpaceName, map[string]bool{"if_not_exists": true}}))
+	if err != nil {
+		return err
+	}
+
+	formatCommand := fmt.Sprintf("box.space.%s:format", benchSpaceName)
+	_, err = tarantoolConnection.Exec(tarantool.Call(formatCommand, [][]map[string]string{
+		{
+			{"name": "key", "type": "string"},
+			{"name": "value", "type": "string"},
+		},
+	}))
+	if err != nil {
+		return err
+	}
+
+	createIndexCommand := fmt.Sprintf("box.space.%s:create_index", benchSpaceName)
+	_, err = tarantoolConnection.Exec(tarantool.Call(createIndexCommand, []interface{}{
+		benchSpacePrimaryIndexName,
+		map[string]interface{}{
+			"parts":         []string{"key"},
+			"if_not_exists": true,
+		},
+	}))
+	return err
+}
+
+func DropSpace(ctx context.BenchCtx, tarantoolConnection *tarantool.Connection) error {
+	checkCommand := fmt.Sprintf("return box.space.%s.index[0].name", benchSpaceName)
+	indexName, err := tarantoolConnection.Exec(tarantool.Eval(checkCommand, []interface{}{}))
+	if err != nil {
+		return err
+	}
+	if reflect.ValueOf(indexName.Data).Index(0).Elem().String() == benchSpacePrimaryIndexName {
+		dropCommand := fmt.Sprintf("box.space.%s:drop", benchSpaceName)
+		_, err := tarantoolConnection.Exec(tarantool.Call(dropCommand, []interface{}{}))
+		return err
+	}
+	return nil
```

Comment:
I think instead of all this you can use functions from this module: https://github.com/tarantool/cartridge-cli/tree/fde10b1d11a254fc976c2430870a9dd3d0e7a0d1/cli/connector

Reply:
Could you please tell what is the reason to use wrapped connection in benchmark tool?
I think that benchmark have to be pure as much as possible.

Could you please tell what do you mean common way and why benchmark operations have to be shared with other cartridge cli code?

Reply:
> Could you please tell what is the reason to use wrapped connection in benchmark tool? I think that benchmark have to be pure as much as possible.
> 
> Could you please tell what do you mean common way and why benchmark operations have to be shared with other cartridge cli code?

Hi, Mihail!

> Could you please tell what is the reason to use wrapped connection in benchmark tool?
> I think that benchmark have to be pure as much as possible.

* If we already have a wrapper around the connector (seems like it is), I prefer to use this wrapper everywhere as an abstraction around connector implementation. And if in the future we want to change connector - it will need to be done in one place.
* If the use of the wrapper affects the benchmark results, then this point can be discussed (but for this discussion we must have measurements with and without the wrapper)

> Could you please tell what do you mean common way and why benchmark operations have to be shared with other cartridge cli code?

About common way:
Now the functions for create/drop spaces are defined so that we can't move them to the connector and use them to create/drop other spaces with different options in other tests/benchmarks, and if we need it, we will write the same code again and again.

I also want to clarify that review is a dialogue and I can be completely wrong, I am always ready to listen the arguments for the opposite point of view and discuss it.

Reply:
> an abstraction around connector implementation

Such abstraction does not contain any useful thing for benchmark tool.

> if in the future we want to change connector

Such words can't be reason — no one knows what will be.

> it will need to be done in one place

Connector for benchmark tool can be freeze forever, because goal of benchmark is not absolute measure, but relative between different tarantool versions, different topology, and different load profile.

>  other tests/benchmarks

There are no other test benchmarks. There is RFC about benchmark tool and it does not contain any thing to share with other cartridge cli code

> this is a part of the cartridge cli code.

This is separate tool of cartridge cli with only shared entry point.

> I also want to clarify that review is a dialogue

Dialogue has price. If you have stable profit plan for this benchmark feature and profit so big, I can speak about abstractions and about future code plans infinite time.

Reply:
> > an abstraction around connector implementation
> 
> Such abstraction does not contain any useful thing for benchmark tool.
> 

)

> > if in the future we want to change connector
> 
> Such words can't be reason — no one knows what will be.
>

I disagree, we can write code that can be simply changed, refactored and extended in the future, or we can write monolithic code that we will rewrite for every change.
 
> > it will need to be done in one place
> 
> Connector for benchmark tool can be freeze forever, because goal of benchmark is not absolute measure, but relative between different tarantool versions, different topology, and different load profile.
> 
> > other tests/benchmarks
> 
> There are no other test benchmarks. There is RFC about benchmark tool and it does not contain any thing to share with other cartridge cli code
> 

Sorry, I don't understand which RFC did you mean?

> > this is a part of the cartridge cli code.
> 
> This is separate tool of cartridge cli with only shared entry point.
> 

Here I can agree with you. So far as `connect`and `bench` are different modules, they shouldn't share their internal code..

> > I also want to clarify that review is a dialogue
> 
> Dialogue has price. If you have stable profit plan for this benchmark feature and profit so big, I can speak about abstractions and about future code plans infinite time.

Sometimes yes, but not always. Initially, when reviewing, I want to know the motivation why it was chosen to do this and not otherwise.

Reply:
>  I can speak about abstractions and about future code plans infinite time.

Hah) You are limited by the time of your life, but in fact, I think you will start repeating yourself after 2-3 hours